src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,136 @@
+            packageSourceKey = packageSourceKey.ToLowerInvariant().Trim();
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    currentNode.IsGlobbing = true;
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    currentNode.Children[c] = new SearchNode();
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (string.IsNullOrEmpty(currentNode.NamespaceId))
+            {
+                currentNode.NamespaceId = namespaceId;
+            }
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+            }
+
+            currentNode.PackageSources.Add(packageSourceKey);
+        }
+
+        /// <summary>
+        /// Get package source names with matching prefix "term" from package namespaces section.
+        /// </summary>
+        /// <param name="term">Search term. Never null. </param>
+        /// <returns>Package source names with matching prefix "term" from package namespaces.</returns>
+        /// <exception cref="ArgumentNullException"> if <paramref name="term"/> is null or empty.</exception>
+        public HashSet<string> GetConfiguredPackageSources(string term)
+        {
+            if (string.IsNullOrWhiteSpace(term))
+            {
+                throw new ArgumentNullException(nameof(term));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            term = term.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+            SearchNode currentNode = _root;
+            SearchNode longestMatchingPrefixNode = null;
+
+            if (currentNode.IsGlobbing)
+            {
+                longestMatchingPrefixNode = currentNode;
+            }
+
+            int i = 0;
+
+            for (; i < term.Length; i++)
+            {
+                char c = term[i];
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    break;
+                }
+
+                currentNode = currentNode.Children[c];
+
+                if (currentNode.IsGlobbing)
+                {
+                    longestMatchingPrefixNode = currentNode;
+                }
+            }
+
+            // Full term match.
+            if (i == term.Length && currentNode.IsValueNode)
+            {
+                return currentNode.PackageSources;
+            }
```

Comment:
There is a bug in this implementation when package name is a prefix of a namespace id stored in the trie. Let's say, if we have following namespaces defined.

- `NuGet.*`  packages from `source1`
- `NuGet.Common.*` packages from `source2`

Let's say we search for `NuGet.Common` package id in the trie then this logic returns empty hashset instead of `source1`.